Add support for Windows

connor4312 opened the pull request:
I had to adjust some linting rules as it looks like `xo` has changed in the years since this module was first written. I've also kept with the `>=4` version constraint, which makes code less pretty. I can update that to >=8 as part of this PR, or a separate PR--let me know if you want that 🙂 

Fixes #5

sindresorhus commented:
> I can update that to >=8 as part of this PR

Yeah, feel free to require Node.js 10 in this PR, and also update the dependencies.

sindresorhus commented on index.js:
```diff
@@ -1,6 +1,7 @@
 'use strict';
 const defaultBrowserId = require('default-browser-id');
 const bundleName = require('bundle-name');
+const windows = require('./windows');
```
You need to add this file to `files` in package.json.

shazron commented:
@connor4312 will you have time to update the PR? if not I can use your code and submit a new PR based on it, with the changes.

connor4312 commented:
I've updated the PR :)

connor4312 commented:
It looks like `com.apple.launchservices.secure.plist` isn't available on the OSX machines that Travis uses: https://travis-ci.com/github/sindresorhus/default-browser/jobs/382419730

sindresorhus commented:
> Yeah, feel free to require Node.js 10 in this PR, and also update the dependencies.

Node.js **10**

sindresorhus commented on package.json:
```diff
@@ -37,13 +38,22 @@
     "default-browser-id": "^2.0.0",
-    "xdg-default-browser": "^2.0.0"
+    "execa": "^0.10.0",
+    "xdg-default-browser": "^2.1.0"
```
Upgrade the dependencies to latest.

sindresorhus commented on windows.js:
```diff
@@ -0,0 +1,46 @@
+const execa = require('execa');
+const util = require('util');
+
+// Windows doesn't have browser IDs in the same way OSX/Linux does--give fake
```
`OSX` => `macOS`

sindresorhus commented on windows.js:
```diff
@@ -0,0 +1,46 @@
+// ones that look real and match the OSX/Linux versions for cross-platform apps.
+const windowsBrowserProgIds = {
+	AppXq0fevzme2pys62n3e0fbqa7peapykr8v: {name: 'Edge', id: 'com.microsoft.edge.old'},
+	MSEdgeDHTML: {name: 'Edge', id: 'com.microsoft.edge'}, // On OSX, its "com.microsoft.edgemac"
```
`it's`

sindresorhus commented on windows.js:
```diff
@@ -0,0 +1,46 @@
+	this.message = message;
+}
+
+util.inherits(UnknownBrowserError, Error);
```
Use normal ES6 subclassing.

sindresorhus commented:
You need to mention Windows support in the readme.

DannyFeliz commented:
@connor4312 👀

connor4312 commented:
Thanks for the ping. Updated dependencies as requested. The update to tsd caused an issue in tests complaining about a lack of `esModuleInterop` for the default import, so I changed that to an `import *`.

connor4312 commented:
Looks like the test failure is caused by `default-browser-id`. Not sure what the deal there is -- the module works on Catalina and Big Sur, maybe Github Actions' macOS setup is set up differently than a standard osx install.

sindresorhus commented on package.json:
```diff
@@ -7,16 +7,17 @@
   },
   "engines": {
-    "node": ">=4"
+    "node": ">=8"
```
You can target Node.js 12 now.

sindresorhus commented on windows.js:
```diff
@@ -0,0 +1,40 @@
+
+	// Execa@0.10 throws if the command fails--no need to check `result.failed`
+
+	const match = /ProgId\s*REG_SZ\s*(\S+)/.exec(result.stdout);
```
Use named capture group.

sindresorhus commented on windows.js:
```diff
@@ -0,0 +1,40 @@
+		'ProgId'
+	]);
+
+	// Execa@0.10 throws if the command fails--no need to check `result.failed`
```
This comment mentions an old Execa version.

sindresorhus commented on windows.js:
```diff
@@ -0,0 +1,40 @@
+
+	const match = /ProgId\s*REG_SZ\s*(\S+)/.exec(result.stdout);
+	if (!match) {
+		throw new UnknownBrowserError(`Cannot find windows browser in stdout: ${JSON.stringify(result.stdout)}`);
```
```suggestion
		throw new UnknownBrowserError(`Cannot find Windows browser in stdout: ${JSON.stringify(result.stdout)}`);
```
